lib/db/schema/patch-018-019.sql

```diff
@@ -0,0 +1,131 @@
+
+CREATE TABLE devices (
+  uid BINARY(16) NOT NULL,
+  id BINARY(16) NOT NULL,
+  sessionTokenId BINARY(32),
+  name VARCHAR(255) NOT NULL,
+  type VARCHAR(16) NOT NULL,
+  createdAt BIGINT UNSIGNED NOT NULL,
+  callbackURL VARCHAR(255),
+  PRIMARY KEY (uid, id)
+) ENGINE=InnoDB;
+
+
+CREATE PROCEDURE `upsertDevice_1` (
+    IN `inUid` BINARY(16),
+    IN `inId` BINARY(16),
+    IN `inSessionTokenId` BINARY(32),
+    IN `inName` VARCHAR(255),
+    IN `inType` VARCHAR(16),
+    IN `inCreatedAt` BIGINT UNSIGNED,
+    IN `inCallbackURL` VARCHAR(255)
+)
+BEGIN
+  INSERT INTO devices(
+    uid,
+    id,
+    sessionTokenId,
+    name,
+    type,
+    createdAt,
+    callbackURL
+  )
+  VALUES (
+    inUid,
+    inId,
+    inSessionTokenId,
+    inName,
+    inType,
+    inCreatedAt,
+    inCallbackURL
+  )
+  ON DUPLICATE KEY UPDATE
+    sessionTokenId = COALESCE(inSessionTokenId, sessionTokenId),
+    name = COALESCE(inName, name),
+    type = COALESCE(inType, type),
+    callbackURL = COALESCE(inCallbackURL, callbackURL);
```

Comment:
I think it would be worth a test that the `ON DUPLICATE KEY UPDATE` logic is correctly triggered here, e.g. doing two calls to `upsertDevice` and confirming that the selected keys are properly overridden.